### DISCUSSION TITLE
Is 2-byte alignment of values guaranteed?

### DISCUSSION THREAD

**Discussion by snej** [erthink/libmdbx#357] [Q&A] [answered]
A minor issue I had with LMDB is that values are not 2-byte aligned. But libmdbx does appear to 2-byte-align values. Is this guaranteed, or is it an implementation detail you don't want clients to depend on? Either way, it would be nice to document that.

**Comment by erthink** [comment, marked as answer]
1. Both LMDB and MDBX have a similar structure of DB pages. Each leaf-page (of a B+tree) is filled with user data (keys and values) end-to-end starting from the end of the page.
   - So basic requirements for get particular alignment are that size of the each key and value must be multiplicity of the desired alignment.
   - It should be clarified that each key+value pair is stored without a gap inside, but it length is rounded up to a 2-bytes. Thus, the beginning of each key is always aligned with the 2-byte boundary, but the value data is always started immediately after the end of key.

2. Then, each branch-page (of a B+tree) is filled similar, but with page-numbers instead of user data. The page number is always 32-bit for MDBX, but depend of `size_t` for LMDB (i.e. 32 or 64 depending of the target platform).
   - So MDBX couldn't provide more than 4-bytes alignment for keys on a branch-pages, but LMDB could provide up to 8-bytes on 64-bit platforms.

3. The next case is large/overflow pages. When some value is too long to be placed on a leaf-page, then it will be place on a overflow/large page (i.e. linear sequence of an one or more pages). In this case value will be placed immediately after page header with corresponding alignment. The size of page header is 20 bytes for MDBX, and 12 or 16 (depending of the platform's bitness) for LMDB.
   - So this also gives 4-byte alignment for MDBX, and up to 8 for LMDB.

4. The next case is filling pages of `dupsort` sub-databases (`MDBX_DUPSORT`, `MDBX_DUPFIXED`, `MDBX_INTEGERDUP`). Briefly, "sorted duplicates" are stored inside nested b+tree.  
   - When only a few values are associated to a key, it is too expensive to use an entire full-size page. Therefore, a short sub-leaf page will be created. So in this case a page header will be added and we have a combination of previously described cases.
   - Otherwise, if a key have enough associated values for an entire full-size page or more, the nested sub-database record will be added. Pages of this nested sub-database have the same format as described above, but multi-values associated with upper-key becomes the nested keys here with zero-sized values. On the other hand, the nested-sub-database record is 8-bytes multiplicity for MDBX and 4/8 bytes for LMDB.
   - So for `dupsort` sub-databases the 4-byte alignment could be guaranteed (but not more) both for LMDB and MDBX if size of all keys and values are multiplicy of 4.

5. The last case is filling leaf-pages of `fixed-length dupsort` sub-databases (`MDBX_DUPFIXED`, `MDBX_INTEGERDUP`). But it is simples - all keys have the same fixed size and all values are zero-sized.
   - So for `fixed-length dupsort` sub-databases the 4-byte alignment could be guaranteed (but not more) both for LMDB and MDBX if size of all keys and values are multiplicy of 4.

-----

Overall:
   - MDBX could provide 4-byte alignment for any sub-database if size of the each key and value are multiplicity 4.
   - MDBX could provide 8-bytes (and more) alignment for non-DUPSORT sub-databases if size of the each key and value are multiplicity of the required alignment.
   - seems LMDB have the same properties.

**Comment by snej** [comment]
Thanks for the info!

> the beginning of each key is always aligned with the 2-byte boundary, but the value data is always started immediately after the end of key.

I checked my code, and the values that need 2-byte alignment always have even-length keys (they're 32-bit ints), which explains why the values are always aligned. For whatever reason, this wasn't happening with LMDB; I got a random mixture of even and odd value addresses.

**Comment by erthink** [comment]
> For whatever reason, this wasn't happening with LMDB; I got a random mixture of even and odd value addresses.

As far as I can remember, I didn't make any changes that caused the MDBX behavior be differ in from the LMDB at this aspect.

On the other hand, I could have done something and forgotten in five years of development. Nonetheless, it is better to ask questions about LMDB to Howard Chu (@hyc).

**Comment by hyc** [comment]
Current LMDB guarantees 2-byte alignment for keys, and values start immediately after the end of the key, just as Leo described. In LMDB 1.0, keys will be padded if necessary, to ensure values start 2-byte aligned. If you have a current LMDB database with purely even-length keys and odd-aligned values, I would be interested to see a test case that reproduces it.
